### ISSUE TITLE
Add ability to call Perl 5 subs/methods in non-array context

### ISSUE THREAD

**Issue by hoelzro** [niner/Inline-Perl5#31]
I noticed that the documentation for Inline::Perl5 doesn't mention context, and digging into the code led me to discover that Perl 5 callables are only ever called in array context, with logic occuring to return the first scalar if exactly one scalar is in the return array.  It may prove handy in the future to allow overriding the context.

**Comment by raiph** [comment]
Hi @niner,

Please consider updating this issue with some sort of roadmap for responding to it -- even if it's just a one liner "This is a dead-end" or "I'm still thinking about this -- let me know about particular modules" or "see these comments here..." or whatever.

I know you added the line in the README saying "Please note that since Perl 6 does not have the same concept of "context", Perl 5 functions are always called in list context.". I've got a vague feeling you concluded it would be too difficult to do anything about this. But it seems very fundamental so I'm wondering if your position is more nuanced. And, anyway, I'd like to have some single link that's your latest "official" position on it.

**Comment by niner** [comment]
I'd absolutely love to find a solution for this!
The best idea I've had so far is to provide a second name for every method/function ending in -scalar. Since Perl 5 doesn't allow "-" in identifiers, those methods shouldn't conflict with any real existing ones. Communicating via method name is also neutral in terms of performance. However, since it is Perl, there actually are ways to declare such methods and call them:
```
> perl -E 'package Foo {}; *{"Foo::foo-scalar"} = sub { say "foo-calar called"; }; my $o = bless {}, "Foo"; my $sub = "foo-scalar"; $o->$sub()'
foo-calar called
```
So this solution is not 100 % safe.

A different way would be to use a special argument, e.g. `$o.foo(Scalar, $foo, $bar)` and let multi dispatch sort it out. This needs some further investigation if it would work in all cases. I've already found out that, it will need an `is default` to resolve the ambiguity with the slurpy arg candidate. I don't know if there's some way to "accidentally" pass a Scalar to a sub. A plain variable is clearly not enough, so there's hope.

Unfortunately I cannot just use a named argument, since those are turned into name => $value pairs to make them work like it was Perl 5 code.

**Comment by niner** [comment]
The more I think about it, the more I like the Scalar solution. If someone could provide some info about where this might break or why it won't break, that would move us much closer to getting it done:

```
> perl6 -e 'multi sub foo(*@a) { @a.say }; multi sub foo (Scalar, *@a) is default { say "scalar!" }; my $bar = "bar"; foo($bar);'
[bar]
> perl6 -e 'multi sub foo(*@a) { @a.say }; multi sub foo (Scalar, *@a) is default { say "scalar!" }; foo(Scalar, "bar");'
scalar!
```

**Comment by lizmat** [comment]
FWIW, in my P5 modules if a function had a different behaviour in scalar context, I have mimicked this by a :scalar named parameter.  And indeed, let MMD take care of execution of the right candidate (using :$scalar!)

**Comment by raiph** [comment]
> I like the Scalar solution

Cute. P6 discloses (deconts) on call, i.e. automatically strips Scalar containers. So there can never be a Scalar container, just a Scalar type object, which is also pretty obscure.

> If someone could provide some info about where this might break or why it won't break, that would move us much closer to getting it done

I'd much rather see jnthn's response to this but in the meantime here's my 2c.

I can only see it breaking for the obvious case in which someone passes a list of arguments starting with the Scalar type object to a P5 call enabled via IP5 without intending it to be interpreted as being in scalar context:

```
my $bar = Scalar;
multi foo (*@a) { say 'array context' }
multi foo (Scalar, *@a) { 'scalar context' }
foo $bar, 42 # Ambiguous call to 'foo(Scalar, Int)'
```

Presumably the only reason a `Scalar` would ever be passed as a first argument is because the arguments are a list of type objects, and `Scalar` just happens to be the first, and the code is written without awareness of IP5's use of this idiom, and the only time it could be a problem is when using IP5 and calling a P5 function with this list of type objects. That's gotta be a pretty rare combination to say the least, one that I can only just about conceive might happen in someone's language exploration code, not in serious or production code. And there'd be a simple workaround of passing arguments so that the first/only one that is or might be Scalar is instead passed inside a list/array:

```
foo (Scalar, Int) # array context
```

Presumably it's possible to eliminate even this theoretical but incredibly unlikely wrinkle by defining another type, say `scalar`, that's expressly intended for this purpose.

That all said, would passing a named argument be as performant, now or in the foreseeable future? If so, I do like that. If you go that route, would there be value to making it be `:P5<scalar>`?

/2c

**Comment by niner** [comment]
I've implemented the Scalar version and it turned out to be surprisingly straight forward. No hidden difficulties at all!

So with commit d89367c8e413a744b9f134555bdaca30cd169d60 this issue is finally fixed :) Thanks to @hoelzro and @raiph for pushing me on this!

**Comment by raiph** [comment]
Does this change mean the "Please note that since Perl 6 does not have the same concept of "context", Perl 5 functions are always called in list context." line in the README should be deleted or changed to something else?

**Comment by niner** [comment]
@raiph indeed it does! Fixed in 4f050194271316b6d74a38653996f82b249f982d
